lettuce/reporters.py

```diff
@@ -165,3 +165,58 @@ def parameter_function(self,i,t,f):
                 + ((grad_u0[2] - grad_u2[0]) * (grad_u0[2] - grad_u2[0]))
             )
         return vorticity.item() * dx**self.lattice.D
+
+
+class SpectrumReporter(GenericStepReporter):
+    """Reports the energy spectrum of the velocity
+    _____
+    NOTES
+    spectrum = simulation.reporters[0].out
+    for i in range(len(spectrum)):
+        k = spectrum[i][1][0]
+        ek = spectrum[i][1][1]
+        plt.loglog(k, ek)
+    plt.show()
+    """
+    _parameter_name = 'Energy spectrum'
+
+    def parameter_function(self,i,t,f):
+        global ek
+        u = self.flow.units.convert_velocity_to_pu(self.lattice.u(f)).cpu().numpy()
+        dx = self.flow.units.convert_length_to_pu(1.0)
+
+        if self.lattice.D == 2:
+            kx = np.fft.fftfreq(self.flow.resolution, d=1 / self.flow.resolution)
+            ky = np.fft.fftfreq(self.flow.resolution, d=1 / self.flow.resolution)
+            kx, ky = np.meshgrid(kx, ky)
+            kk = np.sqrt(kx ** 2 + ky ** 2)
+            norm = dx / self.flow.resolution
+            u0h = np.fft.fftn(u[0] * norm, axes=(0,1))
+            u1h = np.fft.fftn(u[1] * norm, axes=(0,1))
+            ekin = (u0h.real ** 2 + u0h.imag ** 2 + u1h.real ** 2 + u1h.imag ** 2) * .5
+            ek = np.zeros(int(np.max(kk)))
+            k = np.zeros(int(np.max(kk)))
+            for wv in range(int(np.max(kk))):
+                ii, jj = np.where((kk > (wv - 0.5)) & (kk < (wv + 0.5)))
+                ek[wv] = np.sum(ekin[ii, jj])
+                k[wv] = wv
+
+        if self.lattice.D == 3:
+            kx = np.fft.fftfreq(self.flow.resolution, d=1 / self.flow.resolution)
+            ky = np.fft.fftfreq(self.flow.resolution, d=1 / self.flow.resolution)
+            kz = np.fft.fftfreq(self.flow.resolution, d=1 / self.flow.resolution)
+            kx, ky, kz = np.meshgrid(kx, ky, kz)
+            kk = np.sqrt(kx ** 2 + ky ** 2 + kz ** 2)
+            norm = self.flow.resolution * dx ** (-2) * np.sqrt(2 * np.pi)
+            u0h = np.fft.fftn(u[0], axes=(0, 1, 2)) / norm
+            u1h = np.fft.fftn(u[1], axes=(0, 1, 2)) / norm
+            u2h = np.fft.fftn(u[2], axes=(0, 1, 2)) / norm
+            ekin = (u0h.real ** 2 + u1h.real ** 2 + u2h.real ** 2 + u0h.imag ** 2 + u1h.imag ** 2 + u2h.imag ** 2) * .5
+            ek = np.zeros(int(np.max(kk)))
+            k = np.zeros(int(np.max(kk)))
+            for wv in range(int(np.max(kk))):
+                ii, jj, ll = np.where((kk > (wv - 0.5)) & (kk < (wv + 0.5)))
+                ek[wv] = np.sum(ekin[ii, jj, ll])
+                k[wv] = wv
```

Comment:
I don't think you really need a separate case for 3D.
Why not do this through list comprehension? Something like this (not tested):

```
# ... this can go in the init function ...	
dx = self.flow.units.convert_length_to_pu(1.0)
q, *dimensions = f.shape
frequencies = [self.lattice.convert_to_tensor(np.fft.fftfreq(dim, d=dx)) for dim in dimensions]
wavenumbers = torch.stack(torch.meshgrid(*frequencies))
self.wavenorms = torch.norm(wavenumbers, dim=0)   # (your "kk")
self.normalization = dimensions * torch.sqrt(2*np.pi) / dx**2 
self.wavenumbers = torch.arange(int(torch.max(self.wavenorms)))   # your k
# instead of your ii,jj,ll: store as boolean mask of shape [nx, ny, nz, nk]
self.wavemask = (self.wavenorms[...,None] > self.wavenumbers - 0.5) & (self.wavenorms[...,None] <= self.wavenumbers + 0.5) 
	

# ... this can go in the parameter_function ...

uh = torch.stack([torch.fft(u[i], signal_ndim=self.lattice.D) for i in range(self.lattice.D)]) / self.normalization
ekin = torch.sum(0.5 * (uh[...0]**2 + uh[...,1]**2), dim=0)  # real and imaginary parts
ek = ekin[...,None] * self.wavemask  # has shape [nx, ny, nz, nk], now sum over spatial dimensions
ek = ek.sum(dims=torch.arange(self.lattice.D))
```

Reply:
done in separate PR